lib/rjgit.rb

```diff
@@ -55,11 +55,18 @@ def self.object_for_tag(repository, tag)
     end
 
     # http://dev.eclipse.org/mhonarc/lists/jgit-dev/msg00558.html
-    def self.cat_file(repository, blob)
+    def self.cat_file(repository, blob, path = nil)
       jrepo = RJGit.repository_type(repository)
       jblob = RJGit.blob_type(blob)
       # Try to resolve symlinks; return nil otherwise
-      mode = RJGit.get_file_mode(jrepo, jblob)
+      mode = if path
+        last_commit_hash = jrepo.resolve(Constants::HEAD)
```

Comment:
This mirrors our code [here](https://github.com/repotag/rjgit/blob/master/lib/blob.rb#L75), where we are, however, allowing the user to choose the `revstring` as a parameter. The more fundamental issue is the following.

Git file modes are not a property of a blob as such. The same blob can technically have a different file mode in different trees (and therefore in different branches). For `cat_file` purposes, we just want to know if a blob is a symlink, so that we can cat the contents of the linked file, rather than the contents of the symlink-blob. The contents of a symlink-blob will consist of only the path to the linked file. So the question is: relative to what branch do we want to determine whether a blob (in a given tree `path`) is a symlink or not?

I think it makes sense to make `HEAD` (= the 'active' branch) the default here, as `cat_file` is a `Porcelain` method. We are, however, also using this `Porcelain` method to cat the contents of `Blob`s "plumbing style". In that context, 'active' branch doesn't really have a meaning. Here are two options:

* default to`HEAD`, but let the user specify an optional revstring in `Blob#data` and `Porcelain.cat_file`.
* let `Porcelain.cat_file` accept an optional `mode` argument (default: `nil`), and an optional `lookup_mode` argument (default: `true`). In `Blob#data`, call `cat_file` with the blob's mode if the user has set it (or `nil` otherwise) and `lookup_mode = false`.

Reply:
@dometto In the `grep` code path, which is the only use case that requires looking up the mode based on path here, the blob already knows its mode.  I think we can drop the "lookup mode from path" piece in favor of passing the mode into this method.  I've implemented this in ceb3c06a37089a9c3cd00c0fb343eeec4a753584, removing most of the changes I introduced to this method.

Reply:
Good solution, given that `grep` is indeed the only (plumbing) path for which this is an issue. When calling `cat_file` directly (porcelain api), the user will just have to supply the mode themselves. :+1: